File: Python/BFS/NumberofIslands.py
# https://leetcode.com/problems/number-of-islands/

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        visit = set()
        process = deque()
        rows = len(grid)
        cols = len(grid[0])
        count = 0
        def bfs(i, j):
            process.append((i, j))
            while process:
                x, y = process.popleft()
                dire = [[x + 1, y], [x - 1, y], [x, y + 1], [x, y - 1]]
                for xx, yy in dire:
                    if 0 <= xx < rows and 0 <= yy < cols and grid[xx][yy] == '1' and (xx, yy) not in visit:
                        process.append((xx, yy))
                        visit.add((xx, yy))



        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1' and (i, j) not in visit:
                    visit.add((i, j))
                    bfs(i, j)
                    count += 1

        return count

File: C++/DFS/NumberofIslands.cpp
// https://leetcode.com/problems/number-of-islands/

class Solution {
public:
    int m, n;
    
    int numIslands(vector<vector<char>>& grid) {
        int ret = 0;
        m = grid.size();
        n = grid[0].size();
        vector<vector<bool>> visited(m, vector<bool>(n, false));        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(grid, i, j, visited)) {
                    ret += 1;
                }
            }
        }
        
        return ret;
    }
    
    bool dfs(vector<vector<char>>& grid, int i, int j, vector<vector<bool>>& visited) {
        if (i >= m || i < 0 || j >= n || j < 0 || visited[i][j] || grid[i][j] == '0') return false;
        visited[i][j] = true;
        dfs(grid, i + 1, j, visited);
        dfs(grid, i - 1, j, visited);
        dfs(grid, i, j + 1, visited);
        dfs(grid, i, j - 1, visited);
        return true;
    }
};
